compiler+runtime/src/cpp/clojure/string_native.cpp

```diff
@@ -57,6 +62,168 @@ namespace clojure::string_native
     return make_box(util::to_uppercase(s_str));
   }
 
+  static object_ref
+  replace_first_character(object_ref const s, object_ref const match, object_ref const replacement)
+  {
+    auto const s_str(runtime::to_string(s));
+
+    auto const match_char(expect_object<obj::character>(match)->data);
```

Comment:
You can never use `expect_object` when you're receiving unknown input. Instead, you want `try_object`, which will throw if the object type is incorrect. With `expect_object`, you'll just get undefined behavior instead.